### PR TITLE
Rtl8139

### PR DESCRIPTION
**Pull request by paw1a**

### CODE REVIEW

File: moose/drivers/pci.c [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+#include <drivers/pci.h>
+#include <kstdio.h>
+#include <mm/kmalloc.h>
+#include <mm/resource.h>
+#include <fs/posix.h>
+
+#define PCI_CONFIG_ADDRESS 0xcf8
+#define PCI_CONFIG_DATA    0xcfc
+
+// bit 31
```
Comment (Holodome) [inline moose/drivers/pci.c]: -

File: moose/drivers/pci.c [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    for (u8 device_idx = 0; device_idx < 32; device_idx++) {
+        for (u8 func_idx = 0; func_idx < 8; func_idx++) {
+            u32 bdf = BDF(bus_idx, device_idx, func_idx);
+            u16 vendor_id = read_pci_config_u16(bdf, PCI_VENDOR_ID);
+            if (vendor_id == PCI_DEVICE_NOT_EXIST)
+                continue;
+
+            struct pci_device *device = kzalloc(sizeof(*device));
+            if (device == NULL)
```
Comment (Holodome) [inline moose/drivers/pci.c]: memleak

File: moose/include/drivers/rtl8139.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <types.h>
+#include <net/frame.h>
+
+#define RTL8139_VENDOR_ID 0x10ec
+#define RTL8139_DEVICE_ID 0x8139
+
+int init_rtl8139(u8 *mac_addr);
+void rtl8139_send(struct net_frame *frame);
```
Comment (Holodome) [inline moose/include/drivers/rtl8139.h]: void *

File: moose/include/net/frame.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#include <net/eth.h>
+#include <net/arp.h>
+#include <net/icmp.h>
+#include <net/ip.h>
+#include <net/udp.h>
+
+#define MAX_HEADER_SIZE 256
+#define FRAME_BUFFER_SIZE (ETH_PAYLOAD_MAX_SIZE + MAX_HEADER_SIZE)
+
+enum link_type {
```
Comment (Holodome) [inline moose/include/net/frame.h]: kind

File: moose/net/arp.c [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+static struct arp_cache *cache;
+
+int init_arp_cache(void) {
+    cache = kmalloc(sizeof(*cache));
+    if (cache == NULL)
+        return -ENOMEM;
+
+    for (size_t i = 0; i < ARP_CACHE_SIZE; i++) {
+        struct arp_cache_entry *entry = kmalloc(sizeof(*entry));
+        if (entry == NULL)
```
Comment (Holodome) [inline moose/net/arp.c]: memleak

File: moose/net/arp.c [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        list_add(&entry->list, &free_list);
+    }
+
+    init_list_head(&cache->entries);
+    spin_lock_init(&cache->lock);
+
+    return 0;
+}
+
+static int arp_cache_get(u8 *ip_addr, u8 *mac_addr) {
```
Comment (Holodome) [inline moose/net/arp.c]: const

File: moose/net/arp.c [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    memcpy(header->dst_ip, ip_addr, 4);
+
+    frame->inet_type = INET_TYPE_ARP;
+    memcpy(&frame->arp_header, frame->head, sizeof(*header));
+
+    return eth_send_frame(frame, broadcast_mac_addr, ETH_TYPE_ARP);
+}
+
+static int arp_send_reply(struct net_frame *frame) {
+    struct net_frame *reply_frame = get_free_net_frame(SEND_FRAME);
```
Comment (Holodome) [inline moose/net/arp.c]: free

File: moose/net/frame.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    switch (type) {
+    case SEND_FRAME:
+        frame->head = frame->buffer + MAX_HEADER_SIZE;
+        frame->payload = frame->head;
+        break;
+    case RECEIVE_FRAME:
+        frame->head = frame->buffer;
+        break;
+    default:
```
Comment (Holodome) [inline moose/net/frame.c]: --

File: moose/net/frame.c [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    spin_unlock_irqrestore(&lock, flags);
+}
+
+void push_net_frame_head(struct net_frame *frame, size_t offset) {
+    frame->head += offset;
+    frame->size -= offset;
+}
+
+void pull_net_frame_head(struct net_frame *frame, size_t offset) {
+    frame->head -= offset;
```
Comment (Holodome) [inline moose/net/frame.c:110]: expects

File: moose/net/netdaemon.c [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+#define QUEUE_SIZE 32
+
+struct queue_entry {
+    struct net_frame *frame;
+    int is_free;
+};
+
+struct daemon_queue {
+    struct queue_entry entries[QUEUE_SIZE];
```
Comment (Holodome) [inline moose/net/netdaemon.c]: struct net_frame *

File: moose/net/arp.c [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    case ARP_REQUEST:
+        if (memcmp(header->dst_ip, nic.ip_addr, 4) == 0) {
+            arp_cache_add(header->src_ip, header->src_mac);
+            arp_send_reply(frame);
+        }
+        break;
+    case ARP_REPLY:
+        arp_cache_add(header->src_ip, header->src_mac);
+        break;
+    }
```
Comment (Holodome) [inline moose/net/arp.c:208]: default

File: moose/net/frame.c [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    memset(frame, 0, sizeof(*frame));
+    frame->buffer = frame + 1;
+
+    return frame;
+}
+
+int init_net_frames(void) {
+    for (size_t i = 0; i < FREE_FRAMES_COUNT; i++) {
+        struct net_frame *frame = alloc_net_frame();
+        if (frame == NULL)
```
Comment (Holodome) [inline moose/net/frame.c]: leak

File: moose/net/frame.c [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+static LIST_HEAD(free_list);
+
+static spinlock_t lock = SPIN_LOCK_INIT();
+
+static struct net_frame *alloc_net_frame(void) {
+    struct net_frame *frame = kmalloc(sizeof(*frame) + FRAME_BUFFER_SIZE);
+    if (frame == NULL)
+        return NULL;
+
+    memset(frame, 0, sizeof(*frame));
```
Comment (Holodome) [inline moose/net/frame.c]: зачем?

File: moose/net/arp.c [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    if (arp_cache_get(ip_addr, mac_addr) == 0)
+        return 0;
+
+    struct net_frame *frame = get_free_net_frame(SEND_FRAME);
+    if (frame == NULL)
+        return -ENOMEM;
+
+    int err = arp_send_request(frame, ip_addr);
+    if (err)
+        return err;
```
Comment (Holodome) [inline moose/net/arp.c]: leak

File: moose/net/frame.c (hunk 13 above)
Comment (paw1a) [reply moose/net/frame.c]: надо

**Comment by paw1a** [comment]
Add RTL8139 network card driver and basic net protocols from OSI model: 
Link layer: ethernet
Inet layer: arp, ip, icmp
Transport layer: udp
Add PCI driver

**Pull request merged**